subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/fixture/JfrToStacksConverter.java

```diff
@@ -0,0 +1,235 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.performance.fixture;
+
+import com.google.common.base.Charsets;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.Files;
+import org.openjdk.jmc.common.IMCFrame;
+import org.openjdk.jmc.common.IMCStackTrace;
+import org.openjdk.jmc.common.item.IItem;
+import org.openjdk.jmc.common.item.IItemCollection;
+import org.openjdk.jmc.common.item.IMemberAccessor;
+import org.openjdk.jmc.common.item.IType;
+import org.openjdk.jmc.common.item.ItemToolkit;
+import org.openjdk.jmc.common.unit.BinaryPrefix;
+import org.openjdk.jmc.common.unit.IQuantity;
+import org.openjdk.jmc.common.unit.UnitLookup;
+import org.openjdk.jmc.common.util.MemberAccessorToolkit;
+import org.openjdk.jmc.flightrecorder.JfrAttributes;
+import org.openjdk.jmc.flightrecorder.jdk.JdkAttributes;
+import org.openjdk.jmc.flightrecorder.stacktrace.FrameSeparator;
+import org.openjdk.jmc.flightrecorder.stacktrace.FrameSeparator.FrameCategorization;
+import org.openjdk.jmc.flightrecorder.stacktrace.StacktraceFormatToolkit;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * Converts JFR recordings to the collapsed stacks format used by the FlameGraph tool.
+ */
+class JfrToStacksConverter {
+    public void convertToStacks(IItemCollection recording, File targetFile, Options options) {
+        Map<String, Long> foldedStacks = foldStacks(recording, options);
+        writeFoldedStacks(foldedStacks, targetFile);
+    }
+
+    private Map<String, Long> foldStacks(IItemCollection recording, Options options) {
+        StackFolder folder = new StackFolder(options);
+        StreamSupport.stream(recording.spliterator(), false)
+            .flatMap(eventStream -> StreamSupport.stream(eventStream.spliterator(), false))
+            .filter(options.eventType::matches)
+            .filter(event -> getStackTrace(event) != null)
+            .forEach(folder);
+        return folder.getFoldedStacks();
+    }
+
+    private void writeFoldedStacks(Map<String, Long> foldedStacks, File targetFile) {
+        targetFile.getParentFile().mkdirs();
+        try (BufferedWriter writer = Files.newWriter(targetFile, Charsets.UTF_8)) {
+            for (Map.Entry<String, Long> entry : foldedStacks.entrySet()) {
+                writer.write(String.format("%s %d%n", entry.getKey(), entry.getValue()));
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```

Comment:
Any value in adding an additional error message here instead of just rethrowing without an added message?